diff --git a/pjsip/include/pjsua-lib/pjsua.h b/pjsip/include/pjsua-lib/pjsua.h
index 257cf94d2..9143ca5dd 100644
--- a/pjsip/include/pjsua-lib/pjsua.h
+++ b/pjsip/include/pjsua-lib/pjsua.h
@@ -3344,6 +3344,17 @@ PJ_DECL(pj_status_t) pjsua_transport_lis_start( pjsua_transport_id id,
 #   define PJSUA_REG_RETRY_INTERVAL	300
 #endif
 
+/**
+ * Set this to 0 to disable auto re-registration. This is useful for app 
+ * that uses Push Notification and doesn't require auto re-registration.
+ * App can periodically send refresh registration or send it before making 
+ * a call. See 
+ *
+ * Default: 1 (enabled)
+ */
+#ifndef PJSUA_REG_AUTO_REG
+#   define PJSUA_REG_AUTO_REG	1
+#endif
 
 /**
  * This macro specifies the default value for \a contact_rewrite_method
