diff --git a/pjsip/include/pjsua2/account.hpp b/pjsip/include/pjsua2/account.hpp
index c12df4a4c..1a6c5dde6 100644
--- a/pjsip/include/pjsua2/account.hpp
+++ b/pjsip/include/pjsua2/account.hpp
@@ -1160,9 +1160,13 @@ typedef struct AccountIpChangeConfig
 
 public:
     /**
-     * Virtual destructor
+     * Default constructor
      */
-    virtual ~AccountIpChangeConfig()
+    AccountIpChangeConfig() : shutdownTp(true),
+			      hangupCalls(false),
+                              reinviteFlags(PJSUA_CALL_REINIT_MEDIA |
+                                            PJSUA_CALL_UPDATE_CONTACT |
+                                            PJSUA_CALL_UPDATE_VIA)
     {}
 
     /**
