diff --git a/pjmedia/src/pjmedia/sound_port.c b/pjmedia/src/pjmedia/sound_port.c
index 66e2d40d7..c578d14cc 100644
--- a/pjmedia/src/pjmedia/sound_port.c
+++ b/pjmedia/src/pjmedia/sound_port.c
@@ -240,17 +240,37 @@ static pj_status_t start_sound_device( pj_pool_t *pool,
 		     PJ_EBUG);
 
     /* Get device caps */
-    if (snd_port->aud_param.dir & PJMEDIA_DIR_CAPTURE) {
+ //   if ((snd_port->aud_param.dir & PJMEDIA_DIR_CAPTURE) ||
+ //       (snd_port->aud_param.dir & PJMEDIA_DIR_PLAYBACK))
+ //   {
+	//pjmedia_aud_dev_info dev_info;
+ //       pjmedia_aud_dev_index dev_id =
+ //           (snd_port->aud_param.dir & PJMEDIA_DIR_CAPTURE) ?
+ //           snd_port->aud_param.rec_id :
+ //           snd_port->aud_param.play_id;
+
+	//status = pjmedia_aud_dev_get_info(dev_id, &dev_info);
+	//if (status != PJ_SUCCESS)
+	//    return status;
+
+	//snd_port->aud_caps = dev_info.caps;
+ //   }
+
+    if (snd_port->aud_param.dir & PJMEDIA_DIR_CAPTURE)
+    {
 	pjmedia_aud_dev_info dev_info;
+        pjmedia_aud_dev_index dev_id =
+            (snd_port->aud_param.dir & PJMEDIA_DIR_CAPTURE) ?
+            snd_port->aud_param.rec_id :
+            snd_port->aud_param.play_id;
 
-	status = pjmedia_aud_dev_get_info(snd_port->aud_param.rec_id, 
-					  &dev_info);
+	status = pjmedia_aud_dev_get_info(dev_id, &dev_info);
 	if (status != PJ_SUCCESS)
 	    return status;
 
 	snd_port->aud_caps = dev_info.caps;
     } else {
-	snd_port->aud_caps = 0;
+        snd_port->aud_caps = 0;
     }
 
     /* Process EC settings */
@@ -561,10 +581,7 @@ PJ_DEF(pj_status_t) pjmedia_snd_port_set_ec( pjmedia_snd_port *snd_port,
     pjmedia_aud_param prm;
     pj_status_t status;
 
-    /* Sound must be opened in full-duplex mode */
-    PJ_ASSERT_RETURN(snd_port && 
-		     snd_port->dir == PJMEDIA_DIR_CAPTURE_PLAYBACK,
-		     PJ_EINVALIDOP);
+    PJ_ASSERT_RETURN(snd_port, PJ_EINVALIDOP);
 
     /* Determine whether we use device or software EC */
     if ((snd_port->prm_ec_options & PJMEDIA_ECHO_USE_SW_ECHO) == 0 &&
diff --git a/pjsip/src/pjsua-lib/pjsua_aud.c b/pjsip/src/pjsua-lib/pjsua_aud.c
index 6953c524c..fcc920752 100644
--- a/pjsip/src/pjsua-lib/pjsua_aud.c
+++ b/pjsip/src/pjsua-lib/pjsua_aud.c
@@ -1877,14 +1877,47 @@ static pj_status_t open_snd_dev(pjmedia_snd_port_param *param)
     pj_log_push_indent();
 
     if (speaker_only) {
-	status = pjmedia_snd_port_create_player(pjsua_var.snd_pool,
-						-1,
-						param->base.clock_rate,
-						param->base.channel_count,
-						param->base.samples_per_frame,
-						param->base.bits_per_sample, 
-						0,
-						&pjsua_var.snd_port);
+	//status = pjmedia_snd_port_create_player(pjsua_var.snd_pool,
+	//					-1,
+	//					param->base.clock_rate,
+	//					param->base.channel_count,
+	//					param->base.samples_per_frame,
+	//					param->base.bits_per_sample, 
+	//					0,
+	//					&pjsua_var.snd_port);
+
+	pjmedia_snd_port_param cp_param;
+	int dev_id = param->base.play_id;
+
+	/* Normalize dev_id */
+	if (dev_id < 0)
+	    dev_id = PJMEDIA_AUD_DEFAULT_PLAYBACK_DEV;
+
+	pjmedia_snd_port_param_default(&cp_param);
+
+	/*
+	status = pjmedia_aud_dev_default_param(dev_id, &cp_param.base);
+	if (status != PJ_SUCCESS)
+	    return status;
+	    */
+
+	PJ_LOG(4,(THIS_FILE, "opening with orig param flags[%d]", param->base.flags, param->base.ec_enabled));
+	pj_memcpy(&cp_param.base, &param->base, sizeof(cp_param.base));
+	cp_param.base.dir = PJMEDIA_DIR_PLAYBACK;
+	cp_param.base.play_id = dev_id;
+
+	    PJ_LOG(4,(THIS_FILE, "opening with param flags[%d]", cp_param.base.flags, cp_param.base.ec_enabled));
+	/*
+	cp_param.base.clock_rate = param->base.clock_rate;
+	cp_param.base.channel_count = param->base.channel_count;
+	cp_param.base.samples_per_frame = param->base.samples_per_frame;
+	cp_param.base.bits_per_sample = param->base.bits_per_sample;
+	*/
+	cp_param.options = 0;
+	cp_param.ec_options = param->ec_options;
+
+	status = pjmedia_snd_port_create2(pjsua_var.snd_pool, &cp_param,
+					  &pjsua_var.snd_port);
 
     } else {
 	status = pjmedia_snd_port_create2(pjsua_var.snd_pool,
