diff --git a/pjsip/src/pjsua-lib/pjsua_vid.c b/pjsip/src/pjsua-lib/pjsua_vid.c
index cc3dcd122..bf7d1e0d9 100644
--- a/pjsip/src/pjsua-lib/pjsua_vid.c
+++ b/pjsip/src/pjsua-lib/pjsua_vid.c
@@ -1279,16 +1279,6 @@ void pjsua_vid_stop_stream(pjsua_call_media *call_med)
 
     PJ_LOG(4,(THIS_FILE, "Stopping video stream.."));
     pj_log_push_indent();
-    
-    /* Unregister video stream ports (encode+decode) from conference */
-    if (call_med->strm.v.strm_enc_slot != PJSUA_INVALID_ID) {
-	pjsua_vid_conf_remove_port(call_med->strm.v.strm_enc_slot);
-	call_med->strm.v.strm_enc_slot = PJSUA_INVALID_ID;
-    }
-    if (call_med->strm.v.strm_dec_slot != PJSUA_INVALID_ID) {
-	pjsua_vid_conf_remove_port(call_med->strm.v.strm_dec_slot);
-	call_med->strm.v.strm_dec_slot = PJSUA_INVALID_ID;
-    }
 
     pjmedia_vid_stream_send_rtcp_bye(strm);
 
@@ -1327,6 +1317,16 @@ void pjsua_vid_stop_stream(pjsua_call_media *call_med)
     }
     PJSUA_UNLOCK();
 
+    /* Unregister video stream ports (encode+decode) from conference */
+    if (call_med->strm.v.strm_enc_slot != PJSUA_INVALID_ID) {
+	pjsua_vid_conf_remove_port(call_med->strm.v.strm_enc_slot);
+	call_med->strm.v.strm_enc_slot = PJSUA_INVALID_ID;
+    }
+    if (call_med->strm.v.strm_dec_slot != PJSUA_INVALID_ID) {
+	pjsua_vid_conf_remove_port(call_med->strm.v.strm_dec_slot);
+	call_med->strm.v.strm_dec_slot = PJSUA_INVALID_ID;
+    }
+
     if ((call_med->dir & PJMEDIA_DIR_ENCODING) &&
 	(pjmedia_vid_stream_get_stat(strm, &stat) == PJ_SUCCESS) &&
 	stat.tx.pkt)
